ENCD-2304 Reference epigenome matrix

forresttanaka opened the pull request:
* Added a new query_string.js module to libs because the query-string npm module couldn’t return strings useful to us and would reorder query-string parameters. My version of queryString  does encoded-style query-string encoding, and doesn’t reorder query-string key/value pairs. I think we should gradually move to this module so we can eventually drop the query-string npm module.
* I made a new copy of encodedURIComponents in libs/query_encoding.js, but didn’t remove the one in globals.js as this would have required testing all the numerous places we use it. This new version removes the `options` parameter, which we do use in one place but I don’t think it’s needed, but that would have been another development and testing effort unrelated to this ticket. Code going forward should use the one in libs/query_encoding.js so we can eventually remove the one in globals.js.
* With more matrices coming out, I changed the file naming so that the experiment matrix is now in matrix_experiment.js, the audit matrix is in matrix_audit.js, and the reference-epigenome matrix is in matrix_reference_epigenome.js, and I’d expect future ones to follow this pattern. matrix.js from now collects functionality common to multiple matrices. Experiment and audit matrices have not otherwise changed in this branch.
* I had to adjust the view-control registry to include a new optional filtering function because the reference-epigenome matrix looks like an experiment search, but we only want to display the search and report views. We might want to revisit the view-control registry to allow for custom URLs too, or to even do something different.
* Made a small update to <TabPanel> to allow no selected tab by passing null in the selectedTab property.
* The apparent massive changes in matrix.js aren’t massive changes; they’re file name changes. I did maKe small changes to the experiment and audit matrices but not to change their look or function — just to remove some code duplication that the reference-epigenome matrix would have had to have followed if I kept that.

forresttanaka commented:
I found a Firefox-specific bug in that it doesn’t scroll the Reference Epigenome matrix horizontally, so there will be an update for that coming.

keenangraham commented on src/encoded/tests/test_search.py:
```diff
@@ -50,3 +50,24 @@ def test_matrix_view(workbook, testapp):
         'biosample_ontology.classification']['buckets'][0]['biosample_ontology.term_name']['buckets']) > 0
+
+
+def test_reference_epigenome_matrix_view(workbook, testapp):
```
ENCD-4905 has been merged @forresttanaka.

forresttanaka replied:
@keenangraham I moved over the one view test I added, but wasn’t sure if you insisted on a couple more tests or not.

keenangraham replied:
Probably okay. Looks like you are adding back `test_search.py` here though which was deleted in 4905.

forresttanaka commented:
Need to fix merge conflicts in addition to the Firefox bug.

keenangraham commented on src/encoded/tests/test_search_views.py:
```diff
@@ -511,3 +511,27 @@ def test_search_views_audit_view_no_matrix_defined(workbook, testapp):
     assert r.json['description'] == 'Item type does not have requested view defined: {}'
+
+
+def test_search_views_reference_epigenome_matrix_response(workbook, testapp):
```
Formatting nit:

```
def test_search_views_reference_epigenome_matrix_response(workbook, testapp):
    r = testapp.get(
        '/reference-epigenome-matrix/'
        '?type=Experiment&related_series.@type=ReferenceEpigenome'
        '&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus'
    )
    assert 'aggregations' not in r.json
    assert 'facets' in r.json
    assert 'total' in r.json
    assert r.json['@type'] == ['ReferenceEpigenomeMatrix']
    assert r.json['@id'] == (
        '/reference-epigenome-matrix/'
        '?type=Experiment&related_series.@type=ReferenceEpigenome'
        '&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus'
    )
    assert r.json['@context'] == '/terms/'
    assert r.json['notification'] == 'Success'
    assert r.json['title'] == 'Reference Epigenome Matrix'
    assert r.json['total'] > 0
    assert 'filters' in r.json
    assert 'matrix' in r.json
    assert r.json['matrix']['x']['group_by'] == ['assay_title', 'target.label']
    assert r.json['matrix']['x']['label'] == 'Assay'
    assert r.json['matrix']['y']['group_by'] == [
        'biosample_ontology.classification',
        'biosample_ontology.term_name'
    ]
    assert r.json['matrix']['y']['label'] == 'Biosample'
    assert len(
        r.json['matrix']['y']['biosample_ontology.classification']['buckets']
    ) > 0
    assert len(
        r.json['matrix']['y']['biosample_ontology.classification']['buckets'][0]['biosample_ontology.term_name']['buckets']
    ) > 0
```

forresttanaka replied:
Updated to what you have. Yeah, that looks way better.

forresttanaka commented:
I resolved the merge conflict and fixed the Firefox scrolling issue. I also updated the PR demo to the latest, so that’s it for now for pending changes.

https://encd-2304-re-matrix-pr2-fytanaka.demo.encodedcc.org/

forresttanaka commented on src/encoded/static/components/app.js:
```diff
@@ -50,7 +50,8 @@ const portal = {
                 { id: 'sescc', title: 'SE Stem Cell Consortium', url: '/matrix/?type=Experiment&status=released&internal_tags=SESCC', tag: 'collection' },
-                { id: 'reference-epigenomes', title: 'Reference epigenomes', url: '/search/?type=ReferenceEpigenome&status=released', tag: 'collection' },
+                { id: 'reference-epigenomes-human', title: 'Human reference epigenomes', url: '/reference-epigenome-matrix/?type=Experiment&related_series.@type=ReferenceEpigenome&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens', tag: 'collection' },
+                { id: 'reference-epigenomes-mouse', title: 'Mouse reference epigenomes', url: '/reference-epigenome-matrix/?type=Experiment&related_series.@type=ReferenceEpigenome&replicates.library.biosample.donor.organism.scientific_name=Mus+musculus', tag: 'collection' },
```
Reference epigenome organisms seem very bounded to human and mouse, and @hitz asked for this specifically in the “October 31, 2019, 1:56 PM” comment in the ticket.

forresttanaka commented on src/encoded/static/components/matrix_reference_epigenome.js:
```diff
@@ -0,0 +1,715 @@
+            // exists, otherwise don't select a tab.
+            return organisms.includes(selectedOrganisms[0]) ? selectedOrganisms[0] : null;
+        }
+        return null;
```
I think this was to avoid rewriting the URL for what’s likely to be a situation that never happens, but this might be more of a question for @hitz and @gabdank.

forresttanaka commented on src/encoded/static/components/experiment.js:
```diff
@@ -19,12 +20,33 @@ import { BiosampleSummaryString, BiosampleOrganismNames, CollectBiosampleDocs, A
+        views = types.filter(type => type !== results['@type'][0]);
+    }
+    return views;
+};
```
I mentioned this problem in the QA notes in the ticket. My view-controls module has no consistent way to deal with this since I never expected the "type" to be different from what’s actually desired. This needs to be addressed, but I think in another ticket to make the view controls more flexible.

forresttanaka commented on src/encoded/static/components/view_controls.js:
```diff
@@ -107,14 +107,18 @@ class ViewControl {
+        this._registry[type] = { types: new Set(controlTypes.concat(defaultViewControlTypes)) };
+        if (typeFilter) {
+            this._registry[type].typeFilter = typeFilter;
+        }
```
I think this view-control registry probably has to be replaced with something completely different. I had every expectation that ?type=Experiment would always be for experiments. With that now broken, I think the entire model has to change. The type filter is just something to get around that for now.

forresttanaka commented on src/encoded/static/libs/query_string.js:
```diff
@@ -0,0 +1,144 @@
+     * @return {object} Reference to this object for method chaining.
+     */
+    _parse() {
+        const inputQueryElements = this._query.split('&').filter(element => element);
```
Added an explanatory comment.

forresttanaka commented:
New PR demo with changes through 2019-11-12: https://encd-2304-re-matrix-back-pr3-fytanaka.demo.encodedcc.org/

emro commented on src/encoded/static/components/matrix_reference_epigenome.js:
```diff
@@ -0,0 +1,715 @@
+     * Show a scroll indicator depending on current scrolled position.
+     * @param {object} element DOM element to apply shading to
+     */
+    handleScrollIndicator(element) {
```
@forresttanaka did you see this question? It got collapsed by Github so I just wanted to check.

forresttanaka replied:
Discussed with @emro and we’ll leave this as is, as breaking out a small method that uses a React state in a significant way can get difficult to justify.
